### ISSUE TITLE
Error communicating with Digital Ocean

### ISSUE THREAD

**Issue by hazi98** [Gamocosm/Gamocosm#122]
I'm getting a 404 error from Gamocosm after starting a server.
The problem says:
Error communicating with Digital Ocean while starting server: Digital Ocean API HTTP response status not ok: 404: {"id":"not_found","message":"The resource you were accessing could not be found."} . Aborting

This problem has happened twice as now, and I tried to make a new server using a previous snapshot but the problem persists.

**Comment by Raekye** [comment]
Not gonna lie, that may not be a good message. The worst case scenario is both the snapshot and droplet/server somehow got deleted. I did change some related code today, but it should now be *more* conservative/safe handling snapshots. I double checked that the code seems correct, and "losing" a snapshot + droplet "shouldn't be possible" (see below).

On the main server page, under the Digital Ocean panel, after you click refresh cache, do you still see your snapshot?

I just tried I was able to start and stop one of my existing servers without abnormality, so the code is not obviously wrong.

Here's a summary of what goes on:

- [`server_remote.rb`][1] creates the droplet from a snapshot ID if present, or the default/preset image (this does not modify the image used to create the droplet), and saves the Digital Ocean droplet ID to the server
- [`wait_for_starting_server_worker.rb`][2] keeps going until the droplet is `active` (starts `setup_server_worker.rb`)
- [`start_minecraft_worker.rb`][3] used to destroy the saved snapshot, this is after the droplet is successfully created and nothing went wrong during setup (which just involves updating stuff for an existing server). However, it now keeps the old snapshot even after the droplet is created
- [`wait_for_snapshotting_server_worker.rb`][4] only deletes the saved snapshot if:
   1. The snapshot event succeeded
   2. A snapshot for the droplet is found
   3. That snapshot is newer than the saved snapshot

At any point in time, there should always be at least 1 snapshot of the server, even when it's running. When stopping a server, there is a brief period of time where both the old and new snapshot exist. i.e. even when errors/exceptions arise, Gamocosm should never accidentally delete your snapshot or server without a backup

Before the change earlier today, at any time there was always at least one of: the old snapshot, the running server, the new snapshot. Again, should not be possible to delete both the saved snapshot and server (I've also not had any reports of snapshots going missing with the old code)

---
Now let's assume you do have a snapshot/image ID. You should be able to create a brand new server with the same region + size settings. Then go to "Advanced", set the Digital Ocean remote snapshot ID, set the setup stage to 5, and copy the MCSW key from your old server settings. Then you should be able to start a server. However, it sounds like you tried this and it didn't work

[1]: https://github.com/Gamocosm/Gamocosm/blob/master/app/models/server_remote.rb#L54
[2]: https://github.com/Gamocosm/Gamocosm/blob/master/app/workers/wait_for_starting_server_worker.rb#L47
[3]: https://github.com/Gamocosm/Gamocosm/blob/master/app/workers/start_minecraft_worker.rb#L25
[4]: https://github.com/Gamocosm/Gamocosm/blob/master/app/workers/wait_for_snapshotting_server_worker.rb#L72

**Comment by hazi98** [comment]
I'm sorry to tell you that my snapshot is gone, I had no backup whatsoever. I wonder if that is related to trying to delete the server from Gamocosm and trying again to make another server.

**Comment by Raekye** [comment]
Ah, I'm afraid deleting a stopped server will delete the saved snapshot. I should have made it more clear not to delete the old server before starting the new one, and made it more clear on the server deletion page that deleting a stopped server will also delete it's saved snapshot. For what I described above, you could either start the new server first before deleting the old one, or set the saved snapshot id to empty under the "Advanced" tab before deleting the server.

So I'm sorry about your world... hopefully it was relatively new and you didn't lose too much investment in it. I do try to remind people to make their own external backups, but it's not exactly something we think about when we're just playing

---
Now if you still want to help out, can I confirm exactly when did you get a 404 error? It was working fine and yesterday it just randomly appeared, and then you tried creating a new server from the old one? After that point I can definitely see how a 404 would happen, but I'm not aware of how the 404 could appear by itself, that would be a bug I need to find.

Anyways, once again I'm sorry to confirm there would be no way to recover your world from Gamocosm or Digital Ocean's side. I'm especially sorry if it turns out to be a bug on Gamocosm's side, but at the same time, it would be impossible for me to guarantee flawless code. Not exactly comprehensive, but besides the automated tests, I do test directly on gamocosm.com starting and stopping (existing) servers after every code change

**Comment by hazi98** [comment]
Yes, the bug randomly appeared, the server was fine previously until the next time I wanted to start it. Before this event, it was unable to communicate with the server via SSH due to the keys not being copied into the server.

**Comment by Raekye** [comment]
Hmmm ok, that is interesting. This happened on April 14th? I checked the server logs and didn't see any issues, so this might be a bit hard to track down

**Comment by CoWinkKeyDinkInc** [comment]
I'm having this issue right now, also I changed my API key and I can't change it in the usersettings because it thinks I want to change the password

**Comment by Brottweiler** [comment]
Having this issue right now.

A user had started the server, didn't join for 8 minutes so the auto shutdown triggered.

Then the user tried to start the server again and got the 404 error message. 

On digital ocean you can see in the activity tab the process:

![Screenshot_20210728-070613](https://user-images.githubusercontent.com/1613137/127267045-5f82b213-220a-42f0-b42d-8d8a5e29c18b.png)

I can just create a new server but that will give me a new server address to join, and this issue could happen again. 

On Gamocosm's side, right now it appears the server is on because I have a "turn off" button, but I can't turn it off because it's not on.

**Comment by Raekye** [comment]
@CoWinkKeyDinkInc I'm sorry for not getting back to you on it... If you've moved on of course it's understandable, and I apologize for any trouble caused. If for god knows whatever reason you're still using Gamocosm 😅, I can try to look into it now, but I'll need more information.

@Brottweiler Unfortunately there's no way for a user to "reset" the Gamocosm server from that state (would have been nice for these cases, but at least before there weren't any cases of this). I could do it manually if you give me the server ID/URL (probably easiest via gitter or reddit), but it's not really a "quick fix" anymore 😅.

Anyways, the digital ocean activity tab is very helpful for me to look into this. Hopefully I can actually find something this time...

**Comment by Brottweiler** [comment]
>I could do it manually if you give me the server ID/URL (probably easiest via gitter or reddit), but it's not really a "quick fix" anymore 😅.

No problem, creating a new server from the original snapshot was more painless than I first thought. Your instructions from a year back were helpful!
